hooks/pre/12-clear_pulp_content.rb

```diff
@@ -0,0 +1,17 @@
+# See bottom of the script for the command that kicks off the script
+require 'English'
+
+def reset
+  stop_services
+  reset_pulp
+end
+
+if app_value(:clear_pulp_content) && pulp_enabled? && !app_value(:noop)
+  response = ask('Are you sure you want to continue? This will drop all Pulp 2 data including repositories and synced content. [y/n]')
+  if response.downcase != 'y'
+    $stderr.puts '** cancelled **'
+    exit(1)
```

Comment:
We shouldn't exit in `pre`. I think the confirmation should happen in `pre_commit` or perhaps even `pre_validations`.

Reply:
Could you please elaborate on your reasoning here?

Reply:
https://github.com/theforeman/kafo#hooks has a list of hooks. In `pre` all state has been written to disk. That means the answers are going to be persisted. That means you can end up in a weird situation where things are out of sync.

Imagine you call `foreman-installer --my-password-param --clear-pulp-content`. foreman-maintain actually reads some values (including passwords) from the answers. If you cancel, these are now out of sync which means foreman-maintain may be broken.

However, if you exited in `pre_commit` nothing has been written to disk and no changes have been made. In other words, any calls to `exit()` in `pre` is highly suspicious.

This was clearly already wrong in katello-installer, but I'm using these moves as a good chance to review and properly fix behavior.

Reply:
Great, thanks. I've moved that logic to `pre_validations`

Reply:
Can you elaborate why you think `pre_validations` should be used over `pre_commit`?

Reply:
If we are giving the user the option to abort after realizing what `--clear-pulp-content` implies, it's better to do that before starting the interactive wizard.

Reply:
I also debated this internally, but wanted to know your opinion. I also saw the argument that you want to exit early. However, technically you can turn off Pulp during the interactive phase, so the interaction becomes:

* Run installer with argument
* Confirm in dialog
* Turn Pulp off
* Hook doesn't run because `pulp_enabled?` returns false

So the logic I've been following is that if you use answers, `pre_commit` is the right phase.

This is a downside of using a helper like `pulp_enabled?`: it obfuscates that you use answers and what the implications are.

Btw, you can also wonder why we even need an `ask()` at all. The user passed in the argument and you're expected to know what you're doing.

I also realize now that `--reset-data` suffers from the same problem.

Reply:
So, a similar problem exists if you turn off pulp with a command line argument, which in my experience is more common than using the interactive mode (in fact, I've never met a single user that uses interactive mode):

- Run `foreman-installer --clear-pulp-content --no-enable-foreman-proxy-plugin-pulp`
- Regardless of whether the confirmation exists in `pre_commit` or `pre_validations`, the confirmation dialogue doesn't run because `pulp_enabled?` returns false (either due to being disabled in answers, or due to answers being unavailable)
- The reset logic itself doesn't run for the same reason, so the user believes they have disabled the plugin and reset all its data, but in fact the data remains.

With that said, I think there are bigger problems with disabling the plugin after it has been installed: puppet will do nothing to remove the existing pulp server, services, vhosts, mongoDB, etc -- it simply won't go out of its way any longer to ensure they remain present in the correct state either. So as we get closer to removing Pulp2 we need to come up with a strategy to really remove it rather than merely no longer ensuring its presence.

Reply:
Thinking about it more, I suppose that problem (pulp2 removal) is why the upgrade strategy to the point of pulp3 only installations won't be an in place upgrade at all but a data migration to a fresh install.

So I'm therefore not that worried about edge cases involving disabling the pulp2 plugin when it was previously enabled.

Still, as you pointed out, we can't rely on `pulp_enabled?` in `pre_validations` after all, even in the more common case where we are not trying to disable the plugin. So I think the options are to either move the confirmation to `pre_commit` after all as you suggested, or else remove it entirely.

Reply:
You're right that there are other issues. My goal right now is to figure out where things belong and get a good feeling for that. It helps to have these examples.

An in my mind I'm still going back and forth between the two. Either one have benefits and downsides. To me a strong argument about doing it before entering interactive mode is that you can't spend a lot of time answering questions and then end up throwing those away.

Another way to deal with it is to prompt at `pre_validations` and print a warning (probably in `pre`) that roughly says "we know you requested this, but we're not going to do it. That feels like the best compromise.

> puppet will do nothing to remove the existing pulp server, services, vhosts, mongoDB, etc

Well, it probably removes the vhosts since Apache is set up to purge unknown content. I know there are some customers who have turned that off, but they should know that's unsupported for upgrades.

There are Puppet modules which at the top level accept arguments like `service_ensure => stopped` and `package_ensure => absent`. There is a parameter `version` which we can set to `absent`, but then a lot of other things will break because the rest of our module is not designed for it.

Reply:
I like this idea.

I removed the `pulp_enabled?` condition from the confirmation dialogue in `pre_validations`. I'm considering removing the `!app_value(:noop)` as well, since the behavior of the interface should be the same regardless.

I then modified the logic in `pre` to skip the reset and print warnings if `!pulp_enabled?` or `app_value(:noop)`.

Reply:
Regarding confirmation dialogue with `--noop`, I ultimately elected to display a modified version of the dialogue in that that makes it clear nothing will actually happen. This keeps the user experience consistent.